src/main/java/io/cdap/plugin/couchbase/source/JsonObjectToRecordTransformer.java

```diff
@@ -0,0 +1,240 @@
+   * be uniform, though it can be. A SELECT statement that specifies a fixed set of attribute (column) names results
+   * in a uniform set of documents and a SELECT statement that specifies the wild card (*) results in a non-uniform
+   * result set. Wild card statements result in an object with single key that matches bucket name.
+   * <p>
+   * See: <a href="https://docs.couchbase.com/server/current/n1ql/n1ql-intro/queriesandresults.html#queries">
+   * N1QL Queries and Results
+   * </a>
+   *
+   * @param jsonObject result document, possibly in a non-uniform representation.
+   * @return uniform representation of the given result document.
+   */
+  private JsonObject toUniformJsonObject(JsonObject jsonObject) {
+    boolean isWildCardQuery = config.getSelectFields().contains("*");
+    if (isWildCardQuery && jsonObject.containsKey(config.getBucket())) {
+      JsonObject payload = jsonObject.getObject(config.getBucket());
+      if (jsonObject.size() == 1) {
+        return payload;
+      }
+      // Result object can contain other fields for queries such as: SELECT meta(`test-bucket`).id, * from `test-bucket`
+      // Include them as well
+      jsonObject.removeKey(config.getBucket());
+      for (String name : jsonObject.getNames()) {
+        payload.put(name, jsonObject.get(name));
+      }
+      return payload;
+    }
+
+    return jsonObject;
+  }
+
+  private StructuredRecord extractRecord(@Nullable String fieldName, JsonObject object, Schema schema) {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<Schema.Field> fields = Objects.requireNonNull(schema.getFields(), "Schema fields cannot be empty");
+    for (Schema.Field field : fields) {
+      // Use full field name for nested records to construct meaningful errors messages.
+      // Full field names will be in the following format: 'record_field_name.nested_record_field_name'
+      String fullFieldName = fieldName != null ? String.format("%s.%s", fieldName, field.getName()) : field.getName();
+      Schema nonNullableSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable()
+        : field.getSchema();
+      Object value = extractValue(fullFieldName, object.get(field.getName()), nonNullableSchema);
+      builder.set(field.getName(), value);
+    }
+    return builder.build();
+  }
+
+  private Object extractValue(String fieldName, Object object, Schema schema) {
+    if (object == null) {
+      return null;
+    }
+
+    Schema.LogicalType fieldLogicalType = schema.getLogicalType();
+    if (fieldLogicalType != null) {
+      switch (fieldLogicalType) {
+        case DECIMAL:
+          ensureTypeValid(fieldName, object, Number.class);
+          return extractDecimal(fieldName, (Number) object, schema);
+        default:
+          throw new UnexpectedFormatException(String.format("Field '%s' is of unsupported type '%s'", fieldName,
```

Comment:
does this get checked at configure time? If not, this should be validated at configure time

Reply:
Yes. Also, there are [config tests](https://github.com/vladhlinskiy/couchbase/blob/feature/CDAP-16003-couchbase-source-plugin/src/test/java/io/cdap/plugin/couchbase/CouchbaseSourceConfigTest.java#L365) for unsupported types, like the following:
```
  @Test
  public void testValidateSchemaInvalid() {
    Schema schema = Schema.recordOf("invalid-schema", Schema.Field.of("unsupported", Schema.of(Schema.Type.BYTES)));
    CouchbaseSourceConfig config = CouchbaseSourceConfigBuilder.builder(VALID_CONFIG)
      .setSchema(schema.toString())
      .build();

    MockFailureCollector failureCollector = new MockFailureCollector(MOCK_STAGE);
    config.validate(failureCollector);
    assertOutputSchemaValidationFailed(failureCollector, "unsupported");
  }
```